### PR TITLE
Liveblog server side ads: improve element height approximations

### PR DESCRIPTION
**Pull request by domlander**
## What does this change?
 - Update element height approximations following element sizing testing.
 - Add spacing (i.e. padding/margins) to block and element size predictions. 

## Why?
 - The approximations we we using were based on a desktop viewport. We now approximate for a mobile viewport, since the majority of liveblog users are on a mobile device. Note that our server-side predictions for element height on desktop will not be as good.
 - The approximations were informed guesses based on manually measuring the heights of elements in different browsers. Not very scientific. Therefore, we conducted some tests against 108 liveblog pages, where we record each element, our estimate for that elements height, and the elements actual height reported by cypress on a iPhone 12 viewport. By running this test several times whilst changing the inputs, we were able to improve the estimates for all elements, optimising for reducing the average difference between expected height and actual height.
 
The table below shows the final result for each element and how closely we are able to predict their height on mobile devices. Not perfect, but good enough to know where to insert ad slots server side
 
Element | Number of occurrences | Average height diff in px | Average height diff in %
|--|--|--|--|
TextBlockElement | 6562 | 8.7 | 5.26%
BlockquoteBlockElement | 545 | 23.53 | 7.92%
TweetBlockElement | 519 | 108.25 | 21.25%
RichLinkBlockElement | 400 | 11.88 | 9.39%
ImageBlockElement | 368 | 31.68 | 10.88%
YoutubeBlockElement | 25 | 0.76 | 0.32%
VideoYoutubeBlockElement | 22 | 26.84 | 11.77%
SubheadingBlockElement | 14 | 31.57 | 8.79%
InteractiveBlockElement | 3 | 204 | 34%
TableBlockElement | 3 | 14 | 8.67%
CommentBlockElement | 2 | 13.5 | 9%
EmbedBlockElement | 1 | 0 | 0%
GuideAtomBlockElement | 1 | 0 | 0%

Text, Tweet, Image, Blockquote and RichLink element types are the most important to predict, since they appear several times per liveblog page on average.

### CODE REVIEW

File: dotcom-rendering/src/web/lib/liveblogAdSlots.ts [hunk 1]
```diff
@@ -8,12 +8,22 @@ const MAX_INLINE_ADS = 8;
 /**
  * Minimum allowed space between the top of the liveblog container and the highest inline ad.
  */
-const MIN_SPACE_BEFORE_FIRST_AD = 1_000;
+const MIN_SPACE_BEFORE_FIRST_AD = 500;
```
Comment (domlander) [inline dotcom-rendering/src/web/lib/liveblogAdSlots.ts:11]: Currently an advert will appear after the first block. This means it can be side-by-side with the right ad. Ensuring approximately 500px between the top of the liveblog and the first ad means that this will be unlikely to happen

File: dotcom-rendering/src/web/lib/liveblogAdSlots.ts [hunk 2]
```diff
@@ -28,49 +38,57 @@ type BlockElementHeight = {
 
 /**
  * Approximations of the size each block element type will take up on the page in pixels.
- * Predictions are made for desktop, as we prefer to use a value closer to the lower bound,
- * as we would rather display too few ads than too many. Practically, this will mean that we
- * will show the the right frequency of ads on desktop and too few on smaller devices.
+ * Predication are made for mobile viewports, as data suggests that the majority of liveblog page
```
Comment (mxdvl) [inline dotcom-rendering/src/web/lib/liveblogAdSlots.ts:41]:
```suggestion
 * Predictions are made for mobile viewports, as data suggests that the majority of liveblog page
```

**Review by mxdvl — COMMENTED** [review]
so many numbers!

File: dotcom-rendering/src/web/lib/liveblogAdSlots.ts [hunk 3]
```diff
@@ -28,49 +38,57 @@ type BlockElementHeight = {
 
 /**
  * Approximations of the size each block element type will take up on the page in pixels.
- * Predictions are made for desktop, as we prefer to use a value closer to the lower bound,
- * as we would rather display too few ads than too many. Practically, this will mean that we
- * will show the the right frequency of ads on desktop and too few on smaller devices.
+ * Predictions are made for mobile viewports, as data suggests that the majority of liveblog page
```
Comment (arelra) [inline dotcom-rendering/src/web/lib/liveblogAdSlots.ts:41]: Just for my own understanding does this mean we optimise for mobile viewports and not necessarily for desktop? By doing so what is the impact on ad placement for desktop?

File: dotcom-rendering/src/web/lib/liveblogAdSlots.ts [hunk 4]
```diff
@@ -28,49 +38,57 @@ type BlockElementHeight = {
...
+
 const commentElementHeightData: BlockElementHeight = {
 	type: 'model.dotcomrendering.pageElements.CommentBlockElement',
-	elementHeightExcludingText: 64,
+	elementHeightExcludingText: 74,
 	textHeight: {
-		lineHeight: 23,
-		lineLength: 70,
+		lineHeight: 18,
+		lineLength: 28,
```
Comment (arelra) [inline dotcom-rendering/src/web/lib/liveblogAdSlots.ts:58]: I'm finding these numbers a bit hard to parse as separate consts. Would it be possible and fit with the current flow to have a centralised map where they can be looked up? This could then sit in its own file and be exported.

Something like:
```js
{
  'model.dotcomrendering.pageElements.CommentBlockElement' : { 
    elementHeightExcludingText:  74,
    textHeight: {
      lineHeight: 18,
      lineLength: 28
    }
  }
  'model.dotcomrendering.pageElements.EmbedBlockElement' : {
    ...
  }
}
```

File: dotcom-rendering/src/web/lib/liveblogAdSlots.ts (hunk 3 above)
Comment (arelra) [inline dotcom-rendering/src/web/lib/liveblogAdSlots.ts:41]: I wonder if this process is worth documenting in the DCR main repo in `docs/architecture` or `docs/commercial`

**Review by arelra — COMMENTED** [review]
Great to see an approach to server side positioned ads!

File: dotcom-rendering/src/web/lib/liveblogAdSlots.ts (hunk 4 above)
Comment (domlander) [reply dotcom-rendering/src/web/lib/liveblogAdSlots.ts:58]: Yeah agreed. I have made this change.

File: dotcom-rendering/src/web/lib/liveblogAdSlots.ts (hunk 3 above)
Comment (domlander) [reply dotcom-rendering/src/web/lib/liveblogAdSlots.ts:41]: We do optimise for mobile viewports. There are three options:

1, Optimise for mobile viewports

This may mean that we include too many ads on the page, as a mobile viewport height is smaller, however in testing of >100 pages, no desktop page had two ads in the same viewport on desktop. User device type is almost twice as likely to be mobile than it is desktop for liveblog pages, which is the key reason we went with this option.

2, Optimise for desktop viewports

This will mean that we will likely be rendering too few ads on mobile, since the slots are likely to be further apart. Whether we optimise for desktop or mobile, we can ensure that the overall ad-ratio remains consistent to the level it is currently.

3, Create slots for both, populating the only the relevant slots when the page loads, depending on the viewport

This adds more complexity and requires increasing the amount client-side code. However this would improve ad ratio and use a more "correct" amount of ads on both mobile and desktop. I hope we can AB test this in the future.

Comment (domlander) [reply dotcom-rendering/src/web/lib/liveblogAdSlots.ts:61]: I've created a new file in `docs/commercial`

Comment (arelra) [reply dotcom-rendering/src/web/lib/liveblogAdSlots.ts:61]: Thanks for the explanation Dom 👍

File: dotcom-rendering/src/web/lib/liveblogAdSlots.ts (hunk 4 above)
Comment (arelra) [reply dotcom-rendering/src/web/lib/liveblogAdSlots.ts]: Now that this is in place I wonder if there is potential for more simplification.

Could we:

1. use the `CAPIElement._type` property as the key into this map
2. add an optional function to the map to retrieve the appropriate text property for that element
3. `calculateElementHeight` can check if the element has a text function and if so use that

```js
{
  'model.dotcomrendering.pageElements.TextBlockElement' : { 
    heightExcludingText: 0,
    textHeight: {
      lineHeight: 25.5,
      lineLength: 39,
    },
    text: (element) => element.html.replace(/<[^>]+>/g, '')
  },  
  ...
}
```

Which _I think_ would mean `calculateElementSize` doesn't need the switch-case for every type and can call:

```js
calculateElementHeight(elementHeightData[element._type])
```

`TableBlockElement` and the unknown type would need to be handled separately but I think they are the only exceptional cases.

**Review by arelra — APPROVED** [review]
Nice work! 🎉
